include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+     *     // dependent functions are to be evaluated (and differentiated
+     *     // around).
+     *     ad_helper.set_independent_variable(...);
+     *   }
+     *
+     *   // Compute derivatives of the dependent variables
+     *   const auto derivatives = ad_helper.compute_gradients();
+     * @endcode
+     *
+     * The second approach outlined here is more general than the first, and
+     * will work equally well for both taped and tapeless auto-differentiable
+     * numbers.
+     *
+     * @warning ADOL-C does not support the standard threading models used by
+     * deal.II, so this class should \b not be embedded within a multithreaded
+     * function when using ADOL-C number types. It is, however, suitable for use
+     * in both serial and MPI routines.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperBase
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that are used in,
+       * and result from, all computations.
+       */
+      using scalar_type =
+        typename AD::NumberTraits<ScalarType, ADNumberTypeCode>::scalar_type;
+
+      /**
+       * Type definition for the auto-differentiation number type that is used
+       * in all computations.
+       */
+      using ad_type =
+        typename AD::NumberTraits<ScalarType, ADNumberTypeCode>::ad_type;
+
+      /**
+       * @name Constructor / destructor
+       */
+      //@{
+
+      /**
+       * The constructor for the class.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       * that will be used in the definition of the functions that it is
+       * desired to compute the sensitivities of.
+       * @param[in] n_dependent_variables Number of scalar functions to be
+       * defined that will have a sensitivity to the given independent
+       * variables.
+       */
+      ADHelperBase(const unsigned int n_independent_variables,
+                   const unsigned int n_dependent_variables);
+
+      /**
+       * Destructor
+       */
+      virtual ~ADHelperBase();
+
+      //@}
+
+      /**
+       * @name Interrogation of internal information
+       */
+      //@{
+
+      /**
+       * Returns the number of independent variables that this object expects to
+       * work with.
+       */
+      std::size_t
+      n_independent_variables() const;
+
+      /**
+       * Returns the number of dependent variables that this object expects to
+       * operate on.
+       */
+      std::size_t
+      n_dependent_variables() const;
+
+      /**
+       * Prints the status of all queriable data.
+       *
+       * @param[in] stream The output stream to which the values are to be
+       * written.
+       */
+      void
+      print(std::ostream &stream) const;
```

Comment:
can you say that what is actually being printed depends on the AD type used (and specified by the template argument)? At least that's what I think is happening, not having read down to the implementation of the function.

Is this function overloaded in derived classes? If so, should it be `virtual`?